### PR TITLE
[Feature]: Fused CUTLASS GEMM + static FP8 output quant in epilogue

### PR DESCRIPTION
**Pull request by Itssshikhar**
## Purpose
Ref: #25179

Adding a fused kernel that packs scaled GEMM with static FP8 quant in the existing CUTLASS epilogue. This would save two-way trip of BF16 intermediates b/w `cutlass_scaled_mm` & `static_scaled_fp8_quant`.

## Test Plan
Can be reproduced using the following scripts:
```  
python -m pytest tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py
```
```
python -m pytest tests/compile/passes/test_gemm_quant_fusion.py
```
```
  python benchmarks/cutlass_benchmarks/w8a8_benchmarks.py \
    --dtype fp8 \
    --kernels \
      cutlass_fp8_fp8_bf16_scaled_mm_static_fp8_quant_unfused \
      cutlass_fp8_fp8_fp16_scaled_mm_static_fp8_quant_unfused \
      cutlass_fp8_fp8_static_fp8_quant_fused \
    fp8_quant_fusion_bench \
    --shapes \
      128x128x128 \
      32x4096x4096 \
      128x4096x4096 \
      512x4096x4096 \
      128x4096x11008 \
      512x4096x11008

```

## Test Result
microbenchmarking
```
  Results
  | Shape (M,K,N) | Unfused BF16 us | Unfused FP16 us | Fused us | Fused vs best unfused |
  |---|---:|---:|---:|---:|
  | (128,128,128) | 12.379 | 6.849 | 9.780 | 0.700x |
  | (32,4096,4096) | 55.550 | 54.585 | 56.942 | 0.959x |
  | (128,4096,4096) | 125.058 | 125.112 | 124.589 | 1.004x |
  | (512,4096,4096) | 444.419 | 431.686 | 438.997 | 0.983x |
  | (128,4096,11008) | 352.884 | 353.275 | 334.117 | 1.056x |
  | (512,4096,11008) | 1150.137 | 1161.865 | 1111.670 | 1.035x |
```
bigger K like 32k keeps OOM'ing due to memory constraints on my local machine. Ideally it should work.

Kernel-Tests
```
/home/raven/coding-mess/vllm/vllm-env/lib/python3.12/site-packages/pytest_asyncio/plugin.py:208: Pytest
DeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versio
ns of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the d
efault fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture
 loop scopes are: "function", "class", "module", "package", "session"

  warnings.warn(PytestDeprecationWarning(_DEFAULT_FIXTURE_LOOP_SCOPE_UNSET))
============================= test session starts ==============================
platform linux -- Python 3.12.9, pytest-8.3.5, pluggy-1.5.0 -- /home/raven/coding-mess/vllm/vllm-env/bi
n/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/home/raven/coding-me
ss/vllm/.hypothesis/examples'))
rootdir: /home/raven/coding-mess/vllm
configfile: pyproject.toml
plugins: buildkite-test-collector-0.1.9, cov-6.3.0, asyncio-0.24.0, forked-1.6.0, rerunfailures-14.0, t
imeout-2.3.1, shard-0.1.2, mock-3.14.0, subtests-0.14.1, anyio-4.6.2.post1, schemathesis-3.39.15, hydra
-core-1.3.2, hypothesis-6.131.0
asyncio: mode=Mode.STRICT, default_loop_scope=None
collecting ... collected 8 items
Running 8 items in this shard: tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cut
lass_scaled_mm_static_fp8_quant[False-False-32-128-128], tests/kernels/quantization/test_cutlass_scaled
_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant[False-False-128-256-128], tests/kernels/quant
ization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant[False-True-32-128-
128], tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8
_quant[False-True-128-256-128], tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cu
tlass_scaled_mm_static_fp8_quant[True-False-32-128-128], tests/kernels/quantization/test_cutlass_scaled
_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant[True-False-128-256-128], tests/kernels/quanti
zation/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant[True-True-32-128-12
8], tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_q
uant[True-True-128-256-128]

tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant
[False-False-32-128-128] PASSED [ 12%]
tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant
[False-False-128-256-128] PASSED [ 25%]
tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant
[False-True-32-128-128] PASSED [ 37%]
tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant
[False-True-128-256-128] PASSED [ 50%]
tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant
[True-False-32-128-128] PASSED [ 62%]
tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant
[True-False-128-256-128] PASSED [ 75%]
tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant
[True-True-32-128-128] PASSED [ 87%]
tests/kernels/quantization/test_cutlass_scaled_mm_fp8_quant.py::test_cutlass_scaled_mm_static_fp8_quant
[True-True-128-256-128] PASSED [100%]

=============================== warnings summary ===============================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ at
tribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ at
tribute

vllm-env/lib/python3.12/site-packages/torch/jit/_script.py:362: 14 warnings
  /home/raven/coding-mess/vllm/vllm-env/lib/python3.12/site-packages/torch/jit/_script.py:362: Deprecat
ionWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`
.
    warnings.warn(

vllm-env/lib/python3.12/site-packages/schemathesis/generation/coverage.py:305
  /home/raven/coding-mess/vllm/vllm-env/lib/python3.12/site-packages/schemathesis/generation/coverage.p
y:305: DeprecationWarning: jsonschema.exceptions.RefResolutionError is deprecated as of version 4.18.0.
 If you wish to catch potential reference resolution errors, directly catch referencing.exceptions.Unre
solvable.
    ref_error: type[Exception] = jsonschema.RefResolutionError,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================== 8 passed, 17 warnings in 3.08s ========================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```

Compile-pass
```
/home/raven/coding-mess/vllm/vllm-env/lib/python3.12/site-packages/pytest_asyncio/plugin.py:208: Pytest
DeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versio
ns of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the d
efault fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture
 loop scopes are: "function", "class", "module", "package", "session"

  warnings.warn(PytestDeprecationWarning(_DEFAULT_FIXTURE_LOOP_SCOPE_UNSET))
============================= test session starts ==============================
platform linux -- Python 3.12.9, pytest-8.3.5, pluggy-1.5.0 -- /home/raven/coding-mess/vllm/vllm-env/bi
n/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/home/raven/coding-me
ss/vllm/.hypothesis/examples'))
rootdir: /home/raven/coding-mess/vllm
configfile: pyproject.toml
plugins: buildkite-test-collector-0.1.9, cov-6.3.0, asyncio-0.24.0, forked-1.6.0, rerunfailures-14.0, t
imeout-2.3.1, shard-0.1.2, mock-3.14.0, subtests-0.14.1, anyio-4.6.2.post1, schemathesis-3.39.15, hydra
-core-1.3.2, hypothesis-6.131.0
asyncio: mode=Mode.STRICT, default_loop_scope=None
collecting ... collected 3 items
Running 3 items in this shard: tests/compile/passes/test_gemm_quant_fusion.py::test_gemm_quant_fusion_p
ass[TestGemmStaticFP8QuantModel-True], tests/compile/passes/test_gemm_quant_fusion.py::test_gemm_quant_
fusion_pass[TestGemmDynamicFP8QuantModel-False], tests/compile/passes/test_gemm_quant_fusion.py::test_g
emm_quant_fusion_pass[TestGemmPerTokenFP8QuantModel-False]

tests/compile/passes/test_gemm_quant_fusion.py::test_gemm_quant_fusion_pass[TestGemmStaticFP8QuantModel
-True] PASSED [ 33%]
tests/compile/passes/test_gemm_quant_fusion.py::test_gemm_quant_fusion_pass[TestGemmDynamicFP8QuantMode
l-False] PASSED [ 66%]
tests/compile/passes/test_gemm_quant_fusion.py::test_gemm_quant_fusion_pass[TestGemmPerTokenFP8QuantMod
el-False] PASSED [100%]

=============================== warnings summary ===============================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ at
tribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ at
tribute

vllm-env/lib/python3.12/site-packages/astor/op_util.py:92
  /home/raven/coding-mess/vllm/vllm-env/lib/python3.12/site-packages/astor/op_util.py:92: DeprecationWa
rning: ast.Num is deprecated and will be removed in Python 3.14; use ast.Constant instead
    precedence_data = dict((getattr(ast, x, None), z) for x, y, z in op_data)

vllm-env/lib/python3.12/site-packages/torch/jit/_script.py:362: 14 warnings
  /home/raven/coding-mess/vllm/vllm-env/lib/python3.12/site-packages/torch/jit/_script.py:362: Deprecat
ionWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`
.
    warnings.warn(

<frozen importlib._bootstrap_external>:1301
  <frozen importlib._bootstrap_external>:1301: FutureWarning: The cuda.cudart module is deprecated and 
will be removed in a future release, please switch to use the cuda.bindings.runtime module instead.

<frozen importlib._bootstrap_external>:1301
  <frozen importlib._bootstrap_external>:1301: FutureWarning: The cuda.nvrtc module is deprecated and w
ill be removed in a future release, please switch to use the cuda.bindings.nvrtc module instead.

vllm-env/lib/python3.12/site-packages/schemathesis/generation/coverage.py:305
  /home/raven/coding-mess/vllm/vllm-env/lib/python3.12/site-packages/schemathesis/generation/coverage.p
y:305: DeprecationWarning: jsonschema.exceptions.RefResolutionError is deprecated as of version 4.18.0.
 If you wish to catch potential reference resolution errors, directly catch referencing.exceptions.Unre
solvable.
    ref_error: type[Exception] = jsonschema.RefResolutionError,

tests/compile/backend.py:40
  /home/raven/coding-mess/vllm/tests/compile/backend.py:40: PytestCollectionWarning: cannot collect tes
t class 'TestBackend' because it has a __init__ constructor (from: tests/compile/passes/test_gemm_quant
_fusion.py)
    class TestBackend:

tests/compile/passes/test_gemm_quant_fusion.py:50
  /home/raven/coding-mess/vllm/tests/compile/passes/test_gemm_quant_fusion.py:50: PytestCollectionWarni
ng: cannot collect test class 'TestGemmStaticFP8QuantModel' because it has a __init__ constructor (from
: tests/compile/passes/test_gemm_quant_fusion.py)
    class TestGemmStaticFP8QuantModel(_BaseGemmQuantModel):

tests/compile/passes/test_gemm_quant_fusion.py:57
  /home/raven/coding-mess/vllm/tests/compile/passes/test_gemm_quant_fusion.py:57: PytestCollectionWarni
ng: cannot collect test class 'TestGemmDynamicFP8QuantModel' because it has a __init__ constructor (fro
m: tests/compile/passes/test_gemm_quant_fusion.py)
    class TestGemmDynamicFP8QuantModel(_BaseGemmQuantModel):

tests/compile/passes/test_gemm_quant_fusion.py:64
  /home/raven/coding-mess/vllm/tests/compile/passes/test_gemm_quant_fusion.py:64: PytestCollectionWarni
ng: cannot collect test class 'TestGemmPerTokenFP8QuantModel' because it has a __init__ constructor (fr
om: tests/compile/passes/test_gemm_quant_fusion.py)
    class TestGemmPerTokenFP8QuantModel(_BaseGemmQuantModel):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================== 3 passed, 24 warnings in 6.92s ========================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```

### CODE REVIEW

**Comment by Itssshikhar** [comment]
okay, so this was the same question I kept asking myself when trying to e2e test the kernel but couldn't really figure out where it would fit. I went through your suggestions. 

For `Router GEMM` outputs usually go to `softmax` for `top-k` and not really FP8-quant. `MLA` (decode specifically), on the other hand is interesting, because like you said `BMM + o_proj` has the `GEMM + quant` kernel but the obv case is that it's `BMM` not `GEMM` and it usually follows the output with `torch.ops._C.static_scaled_fp8_quant`. This and the fact that I was doing `static per-token` quant here while `MLA` (deepseek's style) uses `per-group/block quant`.

Possible way of taking this forward would be to extend #35792 with a `per-group FP8` for `forward_mqa` (deepseek's usecase). what do you think?
cc @dorhuri123 carlyou
